yjit/src/codegen.rs

```diff
@@ -2835,7 +2836,23 @@ fn gen_setinstancevariable(
             asm.cmp(write_val, Qnil.into());
             asm.jbe(skip_wb);
 
+            // If the receiver is old, we need to run the WB
+            let flags_opnd = asm.load(Opnd::mem(64, recv, RUBY_OFFSET_RBASIC_FLAGS));
+            let anded = asm.and(flags_opnd, (RUBY_FL_PROMOTED as u64).into());
+            asm.cmp(anded, (RUBY_FL_PROMOTED as u64).into());
+            asm.je(run_wb);
+
+            // If the GC is doing incremental marking, we need to run the WB
+            let global_ptr = asm.load(unsafe { &rb_yjit_should_run_wb as *const _ as usize }.into());
+            let must_wb_opnd = asm.load(Opnd::mem(8, global_ptr, 0));
+            asm.cmp(must_wb_opnd, 0.into());
+            asm.jne(run_wb); // Unlikely, CPU will predict not taken by default because it's a forward branch
+
+            asm.jmp(skip_wb); // Unconditional branch, we skip the WB by default
+
             asm_comment!(asm, "write barrier");
+            asm.write_label(run_wb);
+
```

Comment:
Can you add stats counters for the number of times we run the write barriers vs not? Could be useful to see the data on lobsters

I guess I'm slightly worried about increasing the code size here.

Reply:
The most optimal thing to do would be to use side chains to jump to a stub for the write barrier, though I'm not sure if we can do that cleanly. @XrXr what do you think?

Reply:
Seems hard to get right, and also, most of the code size is from the guards, and those would still be there. So doesn't seem like it helps that much.